### PR TITLE
Added private rooms to chat

### PR DESCRIPTION
**Pull request by poradz**
image

### CODE REVIEW

File: app/Chat.php [hunk 1]
```diff
@@ -181,15 +181,60 @@ public static function getRoomsGlobal()
...
+			->select(['name', 'recordid' => 'GL.global_room_id', 'CNT.cnt_new_message', 'CNT.userid'])
 			->from(['GL' => 'u_#__chat_global'])
 			->leftJoin(['CNT' => $subQuery], "CNT.{$roomIdName} = GL.global_room_id AND CNT.userid = {$userId}")
 			->createCommand()->query();
 		$rooms = [];
 		while ($row = $dataReader->read()) {
 			$row['name'] = Language::translate($row['name'], 'Chat');
 			$row['roomType'] = 'global';
-			$row['isPinned'] = true;
+			$row['isPinned'] = isset($row['userid']) ? true : false;
```
Comment (rskrzypczak) [inline app/Chat.php]:
```suggestion
			$row['isPinned'] = isset($row['userid']);
```

File: app/Chat.php [hunk 2]
```diff
@@ -181,15 +181,60 @@ public static function getRoomsGlobal()
...
+		$subQuery = (new Db\Query())
+			->select([
+				'CR.*',
+				'cnt_new_message' => $cntQuery,
+			])
+			->from(['CR' => 'u_yf_chat_rooms_private']);
+		$query = (new Db\Query())
+			->select(['name', 'recordid' => 'GL.private_room_id', 'CNT.cnt_new_message', 'CNT.userid', 'creatorid', 'created'])
+			->from(['GL' => 'u_#__chat_private']);
+		if (\Users_Privileges_Model::getCurrentUserPrivilegesModel()->isAdminUser()) {
```
Comment (rskrzypczak) [inline app/Chat.php]:
```suggestion
		if (User::getUserModel($userId)->isAdmin()) {
```

File: app/Chat.php [hunk 3]
```diff
@@ -891,6 +984,60 @@ public function addToFavorites()
...
+					'creatorid' => $this->userId,
+					'created' => date('Y-m-d H:i:s')
+				]
+			)->execute();
+		Db::getInstance()->createCommand()->insert(
+				static::TABLE_NAME['room']['private'],
+				[
+					'userid' => $this->userId,
+					'last_message' => 0,
+					static::COLUMN_NAME['room']['private'] => Db::getInstance()->getLastInsertID(),
```
Comment (rskrzypczak) [inline app/Chat.php]: getLastInsertID(**??**)

File: public_html/layouts/basic/styles/libraries/_Quasar.scss [hunk 4]
```diff
@@ -8,11 +8,14 @@
 
 	.col,
 	[class*="col-"] {
-		position: unset;
+		position: relative;
 		padding: unset;
 		padding-right: unset;
 		padding-left: unset;
 	}
+	// .q-field__messages.col {
```
Comment (rskrzypczak) [inline public_html/layouts/basic/styles/libraries/_Quasar.scss]: ??

File: install/install_schema/scheme.sql [hunk 5]
```diff
@@ -1377,6 +1377,27 @@ CREATE TABLE `u_yf_chat_messages_group` (
   CONSTRAINT `fk_chat_group_messages` FOREIGN KEY (`groupid`) REFERENCES `vtiger_groups` (`groupid`) ON DELETE CASCADE
 ) ENGINE=InnoDB DEFAULT CHARSET=utf8;
 
+/*Table structure for table `u_yf_chat_messages_private` */
+
+CREATE TABLE `u_yf_chat_messages_private` (
+  `id` int(10) unsigned NOT NULL AUTO_INCREMENT,
+  `privateid` int(10) unsigned NOT NULL,
```
Comment (mariuszkrzaczkowski) [inline install/install_schema/scheme.sql]: add KEY

File: install/install_schema/scheme.sql [hunk 6]
```diff
@@ -1421,6 +1442,16 @@ CREATE TABLE `u_yf_chat_rooms_group` (
   CONSTRAINT `fk_u_yf_chat_rooms_group_users` FOREIGN KEY (`userid`) REFERENCES `vtiger_users` (`id`) ON DELETE CASCADE
 ) ENGINE=InnoDB AUTO_INCREMENT=2 DEFAULT CHARSET=utf8;
 
+/*Table structure for table `u_yf_chat_rooms_private` */
+
+CREATE TABLE `u_yf_chat_rooms_private` (
+  `roomid` int(10) unsigned NOT NULL AUTO_INCREMENT,
+  `userid` int(10) NOT NULL,
```
Comment (mariuszkrzaczkowski) [inline install/install_schema/scheme.sql]: add KEY

File: install/install_schema/scheme.sql [hunk 7]
```diff
@@ -1377,6 +1377,27 @@ CREATE TABLE `u_yf_chat_messages_group` (
   CONSTRAINT `fk_chat_group_messages` FOREIGN KEY (`groupid`) REFERENCES `vtiger_groups` (`groupid`) ON DELETE CASCADE
 ) ENGINE=InnoDB DEFAULT CHARSET=utf8;
 
+/*Table structure for table `u_yf_chat_messages_private` */
+
+CREATE TABLE `u_yf_chat_messages_private` (
+  `id` int(10) unsigned NOT NULL AUTO_INCREMENT,
+  `privateid` int(10) unsigned NOT NULL,
+  `userid` smallint(5) unsigned NOT NULL,
```
Comment (mariuszkrzaczkowski) [inline install/install_schema/scheme.sql]: add KEY

File: install/install_schema/scheme.sql [hunk 8]
```diff
@@ -1421,6 +1442,16 @@ CREATE TABLE `u_yf_chat_rooms_group` (
   CONSTRAINT `fk_u_yf_chat_rooms_group_users` FOREIGN KEY (`userid`) REFERENCES `vtiger_users` (`id`) ON DELETE CASCADE
 ) ENGINE=InnoDB AUTO_INCREMENT=2 DEFAULT CHARSET=utf8;
 
+/*Table structure for table `u_yf_chat_rooms_private` */
+
+CREATE TABLE `u_yf_chat_rooms_private` (
+  `roomid` int(10) unsigned NOT NULL AUTO_INCREMENT,
+  `userid` int(10) NOT NULL,
+  `private_room_id` int(10) unsigned NOT NULL,
```
Comment (mariuszkrzaczkowski) [inline install/install_schema/scheme.sql]: add KEY

File: install/install_schema/scheme.sql [hunk 9]
```diff
@@ -1421,6 +1442,16 @@ CREATE TABLE `u_yf_chat_rooms_group` (
...
+  `last_message` int(10) unsigned DEFAULT NULL,
```
Comment (mariuszkrzaczkowski) [inline install/install_schema/scheme.sql]: add KEY

File: modules/Chat/actions/ChatAjax.php [hunk 10]
```diff
@@ -250,6 +279,40 @@ public function search(App\Request $request)
...
+	}
+
+	/**
+	 * Add participant.
+	 *
+	 * @param \App\Request $request
+	 */
+	public function addParticipant(App\Request $request)
+	{
+		$recordId = $request->getInteger('recordId');
```
Comment (mariuszkrzaczkowski) [inline modules/Chat/actions/ChatAjax.php]: isAdmin

File: modules/Chat/actions/Room.php [hunk 11]
```diff
@@ -103,6 +121,8 @@ private function checkPermissionByRoom(App\Request $request): void
 					throw new \App\Exceptions\NoPermittedToRecord('ERR_NO_PERMISSIONS_FOR_THE_RECORD', 406);
 				}
 				break;
+			case 'private' || 'global':
+				break;
```
Comment (mariuszkrzaczkowski) [inline modules/Chat/actions/Room.php]: ??

**Pull request merged**